File: csharp/db4otools/Statistics.cs
/* Copyright (C) 2004 - 2005  db4objects Inc.  http://www.db4o.com

This file is part of the db4o open source object database.

db4o is free software; you can redistribute it and/or modify it under
the terms of version 2 of the GNU General Public License as published
by the Free Software Foundation and as clarified by db4objects' GPL 
interpretation policy, available at
http://www.db4o.com/about/company/legalpolicies/gplinterpretation/
Alternatively you can write to db4objects, Inc., 1900 S Norfolk Street,
Suite 350, San Mateo, CA 94403, USA.

db4o is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA  02111-1307, USA. */
using System;
using com.db4o.foundation;
using j4o.lang;
using com.db4o;
using com.db4o.ext;
namespace com.db4o.tools
{
/**
     * prints statistics about a database file to System.out.
     * <br>
     * <br>Pass the database file path as an argument.
     * <br>
     * <br>This class is not part of db4o.dll. It is delivered
     * as sourcecode in the path ../com/db4o/tools/<br><br>
     */	public class Statistics
	{
/**
         * the main method that runs the statistics.
         * @param String[] a String array of length 1, with the name
         * of the database file as element 0.
         */		public static void Main(string[] args)
		{
			if (args == null || args.Length != 1)
			{
				Console.WriteLine("Usage: java com.db4o.tools.Statistics <database filename>");
			}
			else
			{
				new Statistics().Run(args[0]);
			}
		}

		public void Run(string filename)
		{
			if (new j4o.io.File(filename).Exists())
			{
				ObjectContainer con1 = null;
				try {
					con1 = Db4o.OpenFile(filename);
					PrintHeader("STATISTICS");
					Console.WriteLine("File: " + filename);
					PrintStats(con1, filename);
					con1.Close();
				}
				catch (Exception e) {
					Console.WriteLine("Statistics failed for file: '" + filename + "'");
					Console.WriteLine(e.Message);
					j4o.lang.JavaSystem.PrintStackTrace(e);
				}
			}
			else
			{
				Console.WriteLine("File not found: '" + filename + "'");
			}
		}

		private void PrintStats(ObjectContainer con, string filename)
		{
			Tree unavailable = new TreeString(REMOVE);
			Tree noConstructor = new TreeString(REMOVE);
			StoredClass[] internalClasses = con.Ext().StoredClasses();
			for (int i1 = 0; i1 < internalClasses.Length; i1++) {
				try {
					Class clazz1 = Class.ForName(internalClasses[i1].GetName());
					try {
						clazz1.NewInstance();
					}
					catch (Exception th) {
						noConstructor = noConstructor.Add(new TreeString(internalClasses[i1].GetName()));
					}
				}
				catch (Exception t) {
					unavailable = unavailable.Add(new TreeString(internalClasses[i1].GetName()));
				}
			}
			unavailable = unavailable.RemoveLike(new TreeString(REMOVE));
			noConstructor = noConstructor.RemoveLike(new TreeString(REMOVE));
			if (unavailable != null)
			{
				PrintHeader("UNAVAILABLE");
				unavailable.Traverse(new StatisticsPrintKey());
			}
			if (noConstructor != null)
			{
				PrintHeader("NO PUBLIC CONSTRUCTOR");
				noConstructor.Traverse(new StatisticsPrintKey());
			}
			PrintHeader("CLASSES");
			Console.WriteLine("Number of objects per class:");
			if (internalClasses.Length > 0)
			{
				Tree all1 = new TreeStringObject(internalClasses[0].GetName(), internalClasses[0]);
				for (int i1 = 1; i1 < internalClasses.Length; i1++) {
					all1 = all1.Add(new TreeStringObject(internalClasses[i1].GetName(), internalClasses[i1]));
				}
				all1.Traverse(new StatisticsPrintNodes());
			}
			PrintHeader("SUMMARY");
			Console.WriteLine("File: " + filename);
			Console.WriteLine("Stored classes: " + internalClasses.Length);
			if (unavailable != null)
			{
				Console.WriteLine("Unavailable classes: " + unavailable.Size());
			}
			if (noConstructor != null)
			{
				Console.WriteLine("Classes without public constructors: " + noConstructor.Size());
			}
			Console.WriteLine("Total number of objects: " + (ids.Size() - 1));
		}

		private void PrintHeader(string str)
		{
			int starcount = (39 - str.Length) / 2;
			string stars = "";
			for (int i1 = 0; i1 < starcount; i1++) {
				stars += "*";
			}
			Console.WriteLine("\n\n" + stars + " " + str + " " + stars);
		}

		static internal TreeInt ids = new TreeInt(0);

		private static string REMOVE = "XXxxREMOVExxXX";

	}
	internal class StatisticsPrintKey : Visitor4
	{
		public void Visit(object obj)
		{
			Console.WriteLine(((TreeString)obj).i_key);
		}

	}
	internal class StatisticsPrintNodes : Visitor4
	{
		public void Visit(object obj)
		{
			TreeStringObject node = (TreeStringObject)obj;
			long[] newIDs = ((StoredClass)node.i_object).GetIDs();
			for (int j = 0; j < newIDs.Length; j++) {
				if (Statistics.ids.Find(new TreeInt((int)newIDs[j])) == null)
				{
					Statistics.ids = (TreeInt)Statistics.ids.Add(new TreeInt((int)newIDs[j]));
				}
			}
			Console.WriteLine(node.i_key + ": " + newIDs.Length);
		}

	}
}

File: java/src/com/db4o/tools/Statistics.java
/* Copyright (C) 2004 - 2005  db4objects Inc.  http://www.db4o.com

This file is part of the db4o open source object database.

db4o is free software; you can redistribute it and/or modify it under
the terms of version 2 of the GNU General Public License as published
by the Free Software Foundation and as clarified by db4objects' GPL 
interpretation policy, available at
http://www.db4o.com/about/company/legalpolicies/gplinterpretation/
Alternatively you can write to db4objects, Inc., 1900 S Norfolk Street,
Suite 350, San Mateo, CA 94403, USA.

db4o is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA  02111-1307, USA. */
package com.db4o.tools;

import com.db4o.*;
import com.db4o.ext.*;
import com.db4o.foundation.*;

/**
 * prints statistics about a database file to System.out.
 * <br><br>Pass the database file path as an argument.
 * <br><br><b>This class is not part of db4o.jar!</b><br>
 * It is delivered as sourcecode in the
 * path ../com/db4o/tools/<br><br>
 */
public class Statistics {

	/**
	 * the main method that runs the statistics.
	 * @param String[] a String array of length 1, with the name of the database
	 * file as element 0.
	 */
	public static void main(String[] args) {
		Db4o.configure().messageLevel(-1);
		if (args == null || args.length != 1) {
			System.out.println("Usage: java com.db4o.tools.Statistics <database filename>");
		} else {
			new Statistics().run(args[0]);
		}
	}

	public void run(String filename) {
		if (new java.io.File(filename).exists()) {
			ObjectContainer con = null;
			try {
				con = Db4o.openFile(filename);
				printHeader("STATISTICS");
				System.out.println("File: " + filename);
				printStats(con, filename);
				con.close();
			} catch (Exception e) {
				System.out.println("Statistics failed for file: '" + filename + "'");
				System.out.println(e.getMessage());
				e.printStackTrace();
			}
		} else {
			System.out.println("File not found: '" + filename + "'");
		}
	}

	private void printStats(ObjectContainer con, String filename) {

		Tree unavailable = new TreeString(REMOVE);
		Tree noConstructor = new TreeString(REMOVE);

		final TreeInt[] ids = { new TreeInt(0)};
		// one element too many, substract one in the end

		StoredClass[] internalClasses = con.ext().storedClasses();
		for (int i = 0; i < internalClasses.length; i++) {
			try {
				Class clazz = Class.forName(internalClasses[i].getName());
				try {
					clazz.newInstance();
				} catch (Throwable th) {
					noConstructor =
						noConstructor.add(new TreeString(internalClasses[i].getName()));
				}
			} catch (Throwable t) {
				unavailable = unavailable.add(new TreeString(internalClasses[i].getName()));
			}
		}
		unavailable = unavailable.removeLike(new TreeString(REMOVE));
		noConstructor = noConstructor.removeLike(new TreeString(REMOVE));
		if (unavailable != null) {
			printHeader("UNAVAILABLE");
			unavailable.traverse(new Visitor4() {
				public void visit(Object obj) {
					System.out.println(((TreeString) obj).i_key);
				}
			});
		}
		if (noConstructor != null) {
			printHeader("NO PUBLIC CONSTRUCTOR");
			noConstructor.traverse(new Visitor4() {
				public void visit(Object obj) {
					System.out.println(((TreeString) obj).i_key);
				}
			});
		}

		printHeader("CLASSES");
		System.out.println("Number of objects per class:");

		if (internalClasses.length > 0) {
			Tree all = new TreeStringObject(internalClasses[0].getName(), internalClasses[0]);
			for (int i = 1; i < internalClasses.length; i++) {
				all =
					all.add(
						new TreeStringObject(internalClasses[i].getName(), internalClasses[i]));
			}
			all.traverse(new Visitor4() {
				public void visit(Object obj) {
					TreeStringObject node = (TreeStringObject) obj;
					long[] newIDs = ((StoredClass)node.i_object).getIDs();
					for (int j = 0; j < newIDs.length; j++) {
						if (ids[0].find(new TreeInt((int) newIDs[j])) == null) {
							ids[0] = (TreeInt) ids[0].add(new TreeInt((int) newIDs[j]));
						}
					}
					System.out.println(node.i_key + ": " + newIDs.length);
				}
			});

		}

		printHeader("SUMMARY");
		System.out.println("File: " + filename);
		System.out.println("Stored classes: " + internalClasses.length);
		if (unavailable != null) {
			System.out.println("Unavailable classes: " + unavailable.size());
		}
		if (noConstructor != null) {
			System.out.println("Classes without public constructors: " + noConstructor.size());
		}
		System.out.println("Total number of objects: " + (ids[0].size() - 1));
	}
	
	private void printHeader(String str){
		int stars = (39 - str.length()) / 2;
		System.out.println("\n");
		for (int i = 0; i < stars; i++) {
			System.out.print("*");
		}
		System.out.print(" " + str + " ");
		for (int i = 0; i < stars; i++) {
			System.out.print("*");
		}
		System.out.println();
	}

	private static final String REMOVE = "XXxxREMOVExxXX";

}
